File: packages/foundation/moodiary_rust/rust/src/api/http_server.rs
use anyhow::Result;
use flutter_rust_bridge::{DartFnFuture, frb};
use std::sync::Arc;

use crate::api::http::KeyValue;

pub use crate::http::server::{HttpServerRequest, HttpServerResponse};

#[frb(mirror(HttpServerRequest))]
pub struct _HttpServerRequest {
    pub method: String,
    pub path: String,
    pub query: Vec<KeyValue>,
    pub headers: Vec<KeyValue>,
    pub body: Vec<u8>,
    pub body_file_path: Option<String>,
}

#[frb(mirror(HttpServerResponse))]
pub struct _HttpServerResponse {
    pub status: u16,
    pub headers: Vec<KeyValue>,
    pub body: Vec<u8>,
    pub body_file_path: Option<String>,
}

fn fallback_response(error: anyhow::Error) -> HttpServerResponse {
    HttpServerResponse {
        status: 500,
        headers: vec![KeyValue {
            key: "content-type".to_owned(),
            value: "text/plain; charset=utf-8".to_owned(),
        }],
        body: error.to_string().into_bytes(),
        body_file_path: None,
    }
}

#[frb(opaque)]
pub struct HttpServer {
    inner: crate::http::server::HttpServer,
}

impl HttpServer {
    pub async fn start(
        preferred_port: u16,
        loopback_only: bool,
        spool_dir: String,
        // 不能叫 handler：与 FRB 生成代码里的内部调度器字段同名冲突
        on_request: impl Fn(HttpServerRequest) -> DartFnFuture<Result<HttpServerResponse>>
        + Send
        + Sync
        + 'static,
        on_body_progress: impl Fn(i64, i64) -> DartFnFuture<Result<()>> + Send + Sync + 'static,
    ) -> Result<HttpServer> {
        let handler: crate::http::server::HandlerFn = Arc::new(move |req| {
            let call = on_request(req);
            Box::pin(async move { call.await.unwrap_or_else(fallback_response) })
        });
        let progress: crate::http::server::ProgressFn = Arc::new(move |received, total| {
            let call = on_body_progress(received, total);
            Box::pin(async move {
                let _ = call.await;
            })
        });
        Ok(HttpServer {
            inner: crate::http::server::HttpServer::start(
                preferred_port,
                loopback_only,
                spool_dir,
                handler,
                progress,
            )
            .await?,
        })
    }

    #[frb(sync)]
    pub fn port(&self) -> u16 {
        self.inner.port()
    }

    #[frb(sync)]
    pub fn stop(&mut self) {
        self.inner.stop();
    }
}

File: packages/core/moodiary_http/lib/src/http_server.dart
import 'dart:convert';
import 'dart:typed_data';

class HttpServerRequest {
  final String method;

  final String path;

  final Map<String, String> query;

  final Map<String, String> headers;

  final Uint8List body;

  final String? bodyFilePath;

  const HttpServerRequest({
    required this.method,
    required this.path,
    this.query = const {},
    this.headers = const {},
    required this.body,
    this.bodyFilePath,
  });
}

class HttpServerResponse {
  final int statusCode;
  final Map<String, String> headers;
  final Uint8List body;
  final String? bodyFilePath;

  HttpServerResponse(
    this.statusCode, {
    this.headers = const {},
    Uint8List? body,
    this.bodyFilePath,
  }) : body = body ?? Uint8List(0);

  factory HttpServerResponse.ok(List<int> body, {String? contentType}) =>
      HttpServerResponse(
        200,
        headers: {'content-type': ?contentType},
        body: .fromList(body),
      );

  factory HttpServerResponse.json(Object? value) => HttpServerResponse.ok(
    utf8.encode(jsonEncode(value)),
    contentType: 'application/json; charset=utf-8',
  );

  factory HttpServerResponse.text(int statusCode, String message) =>
      HttpServerResponse(
        statusCode,
        headers: const {'content-type': 'text/plain; charset=utf-8'},
        body: .fromList(utf8.encode(message)),
      );

  factory HttpServerResponse.file(
    String path, {
    required String contentType,
    Map<String, String> headers = const {},
  }) => HttpServerResponse(
    200,
    headers: {...headers, 'content-type': contentType},
    bodyFilePath: path,
  );

  factory HttpServerResponse.notFound() =>
      HttpServerResponse.text(404, 'not found');
}

typedef HttpServerHandler = Future<HttpServerResponse> Function(
  HttpServerRequest request,
);

abstract class IHttpServer {
  IHttpServer();

  Future<void> start({
    required HttpServerHandler handler,
    int preferredPort = 0,
    bool loopbackOnly = false,
    String? spoolDir,
    void Function(int received, int? total)? onBodyProgress,
  });

  int get port;

  bool get isRunning;

  Future<void> stop();
}
